Reset password with urls

juanmbellini opened the pull request:
Closes #102

juanmbellini committed:
Add ResetPasswordToken model

- Include Database migration
- Include JPA mapping

juanmbellini committed:
Add interfaces to operate with reset password tokens

- Add a ResetPasswordTokenDao with create and find by nonce methods
- Add methods to UserService to start the process of password reset (i.e create a token, send email), and finish the process (change password, given a token)

juanmbellini commented:
@julianrod94 Faltaría implementar. Mira los files changes para ver donde están los TODOs
Branchea de esta branch y haceme una PR con lo que agregues
Yo después termino de implementar los endpoints en la capa web

julianrod94 committed:
Change Mail template and make resetPassword return the reset url

lipusal commented on paw-webapp/webapp/src/main/java/ar/edu/itba/paw/webapp/mail/MailServiceImpl.java:
```diff
@@ -52,9 +52,9 @@ public void prepare(MimeMessage mimeMessage) throws Exception {
-                        + ", \nYour new password is: "
-                        + newPassword
-                        + "\nPlease log in with your new password, and if you wish, head to your profile page and change your password to something easier to remember.");
+                        + ", \nThis email was sent to you because someone requested a password reset on your account. \n\n Visit the following URL to set a new password: "
```
`someone` queda re mal, es como que un flaco cualquiera vino y pidió que te reseteen la cuenta (más allá de que se pueda hacer o no en nuestra página, queda mal decirlo así). Yo pondría algo como `This email was sent to you after a request to reset your account password.  Visit...`

julianrod94 committed:
move mailService to service and its deppendencies

julianrod94 committed:
needs testing

juanmbellini commented on paw-webapp/service/pom.xml:
```diff
@@ -43,10 +43,36 @@
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
```
Solo hace falta la primera dependencia para tener logging (```slf4j-api```).

juanmbellini commented on paw-webapp/persistence/src/main/java/ar/edu/itba/paw/webapp/persistence/ResetPasswordTokenHibernateDao.java:
```diff
@@ -1,27 +1,50 @@
-        // TODO: implement
-        return null;
+
+        TypedQuery<ResetPasswordToken> baseQuery = em.createQuery("FROM reset_password_tokens AS R where R.nonce = :nonce", ResetPasswordToken.class);
```
No hay que usar el nombre de la tabla, sino el nombre de la clase. Te lo cambio

juanmbellini replied:
A todo esto, hay un método de ```DaoHelper``` que te hace todo esto de una. Te lo cambio para que lo use

juanmbellini committed:
Mini refactor to ResetPasswordTokenHibernateDao
- Use DaoHelper method
- Throw exception after max iters was reached

juanmbellini committed:
Mini refactor to service latyer stuff
- Make a var be private
- Remove used logging dependencies
- Use message format
- Make MailServiceImpl logger be final and static

juanmbellini committed:
Improve the email sent

juanmbellini committed:
Finish implementing the UserService method to reset password
- Add missing validations

juanmbellini committed:
Finish implementing the web layer for reset password
- Add endpoints
- Add security stuff to let the request pass
- Fix component scan of web config (remove mail subpackge which does not exist any more)

lipusal commented on paw-webapp/model/src/main/java/ar/edu/itba/paw/webapp/model/ResetPasswordToken.java:
```diff
@@ -0,0 +1,132 @@
+     * The id of this token.
+     */
+    @Id
+    @SequenceGenerator(name = "shelves_seq", sequenceName = "shelves_id_seq", allocationSize = 1)
```
Hay que hacer una nueva secuencia, no usar la de shelves, super turbio eso

juanmbellini replied:
cierto!
jajaaj copy paste

lipusal commented on paw-webapp/persistence/src/main/resources/migration/V1_201712120708__Add_Reset_Password_Tokens.sql:
```diff
@@ -0,0 +1,11 @@
+);
+
+CREATE UNIQUE INDEX reset_password_tokens_nonce_unique_index
+  ON reset_password_tokens (nonce);
```
Creá la secuencia acá y ponele nombre, y usá ese nombre en el modelo

lipusal commented on paw-webapp/service/src/main/java/ar/edu/itba/paw/webapp/service/MailServiceImpl.java:
```diff
@@ -37,33 +37,35 @@ private InternetAddress from() {
-                        + ", \nYour new password is: "
-                        + newPassword
-                        + "\nPlease log in with your new password, and if you wish, head to your profile page and change your password to something easier to remember.");
+                        + user.getUsername() + ", \n" +
```
Hay que ver si esto se manda como HTML o texto. Si como texto plano, hay que ver que gmail auto-reconozca el link. Si como HTML, hay que reemplazar los `\n` con `<br />`

juanmbellini replied:
Yo supongo que se aguanta texto plano. Pero sino, ni idea

lipusal commented on paw-webapp/service/src/main/java/ar/edu/itba/paw/webapp/service/MailServiceImpl.java:
```diff
@@ -75,19 +77,20 @@ public void prepare(MimeMessage mimeMessage) throws Exception {
+                mimeMessage.setText("Dear " + user.getUsername() + ", \n" +
+                        "Your password was changed.\n" +
+                        "If it wasn't you, please reset it clicking on \"Forgot Password?\" in the log in page." +
+                        "Your sincerely, \n" +
```
Your**s** sincerely

lipusal commented on paw-webapp/service/src/main/java/ar/edu/itba/paw/webapp/service/UserServiceImpl.java:
```diff
@@ -338,8 +336,44 @@ public void unFollowUser(long unFollowedId, User unFollower) {
 
-    // =====================
+    @Override
+    public void requireResetPassword(long userId, String urlTemplate) {
```
Habría que validar que `urlTemplate` tenga el `{0}`, o el format fallaría y le mandamos un `4xx` al cliente?

juanmbellini replied:
mmmm, podría hacrlo en el controller y pasarle la url armada ya

lipusal commented on paw-webapp/service/src/main/java/ar/edu/itba/paw/webapp/service/UserServiceImpl.java:
```diff
@@ -338,8 +336,44 @@ public void unFollowUser(long unFollowedId, User unFollower) {
+        try {
+            //noinspection ConstantConditions  Won't be null as an exception would be thrown
+            stringNonce = new String(Base64Utils.decodeFromUrlSafeString(resetPasswordTokenNonce),
+                    StandardCharsets.UTF_8);
```
El `nonce` es un número, para qué decodificar base 64? No va a tener nada de URL tampoco

juanmbellini replied:
Te viene como string (para que no quede una url con un - en el medio)

juanmbellini replied:
O sea, me había faltado encodearlo cuando lo mandaba en el link del email, pero ya está

juanmbellini committed:
Fix sequence for reset password token

- There was a copy/paste error

juanmbellini committed:
Add missing @Repository annotation to token dao

juanmbellini committed:
Fix typo in generated email

juanmbellini committed:
Add validation of template

- Check if "{0}" is included

juanmbellini committed:
Add missing encoding in the requestReset method
- Was sending url without encoding

juanmbellini committed:
Enable Async emails

- Do not wait confirmation of email sent (makes request too slow)
- Fix some mini typos

juanmbellini committed:
Fix mini bug to check existence of nonce

- Was checking != null instead off == null

juanmbellini committed:
Add missing json validation

- If json with password is not included, throw MissingJsonException

juanmbellini committed:
Fix RegexRequestMatchers for reset password mechanism

- Query params must be included!!

juanmbellini committed:
Final fix to email template sent to users

juanmbellini commented:
@lipusal Esto ya está, quedo joSHa

juanmbellini commented on paw-webapp/model/src/main/java/ar/edu/itba/paw/webapp/model/ResetPasswordToken.java:
```diff
@@ -0,0 +1,132 @@
+     * The id of this token.
+     */
+    @Id
+    @SequenceGenerator(name = "reset_password_tokens_seq", sequenceName = "reset_password_tokens_id_seq", allocationSize = 1)
```
Por más que no los haya agregado en la migration, estos son los nombres que te pone postgres a la hora de crear una sequence usando bigserial o serial, así que no me lo critiques @lipusal  ajajaj

juanmbellini committed:
Mini bug fix: template contains {0} validation was not well done

lipusal commented on paw-webapp/service/src/main/java/ar/edu/itba/paw/webapp/service/MailServiceImpl.java:
```diff
@@ -37,57 +37,60 @@ private InternetAddress from() {
-                        + " the Log In page clicking on \"Forgot Password?\"");
+                mimeMessage.setText("Dear " + user.getUsername() + ", \n" +
+                        "Your password was changed.\n" +
+                        "If it wasn't you, please reset it clicking on \"Forgot Password?\" in the log in page.\n\n" +
```
Agregá `Or if you're already logged in, by clicking on the "Change Password" button in your profile page.`

lipusal reviewed (commented):
Agregale un mini cambio please

lipusal committed:
Change mail message

lipusal committed:
Merge branch 'master' into feature/reset-pw-url
